ui/dashboards/src/components/Variables/VariableEditorForm.tsx

```diff
@@ -0,0 +1,306 @@
+  Tooltip,
+} from '@mui/material';
+import { useImmer } from 'use-immer';
+import { useListPluginMetadata, PluginSpecEditor, usePluginRegistry } from '@perses-dev/plugin-system';
+import { VariableDefinition, TextVariableDefinition, ListVariableDefinition } from '@perses-dev/core';
+import React from 'react';
+
+const VARIABLE_TYPES = ['ListVariable', 'TextVariable'] as const;
+
+function getInitialState(initialVariableDefinition: VariableDefinition) {
+  const textVariableFields = {
+    value: (initialVariableDefinition as TextVariableDefinition).spec.value ?? '',
+  };
+
+  const listVariableFields = {
+    allowMultiple: false,
+    allowAll: false,
+    plugin: {
+      kind: '',
+      spec: {},
+    },
+  };
+  if (initialVariableDefinition.kind === 'ListVariable') {
+    listVariableFields.allowMultiple = initialVariableDefinition.spec.allow_all_value ?? false;
+    listVariableFields.allowAll = initialVariableDefinition.spec.allow_all_value ?? false;
+    listVariableFields.plugin = initialVariableDefinition.spec.plugin;
+  }
+
+  return {
+    name: initialVariableDefinition.spec.name,
+    label: initialVariableDefinition.spec.display?.label,
+    kind: initialVariableDefinition.kind,
+    description: '',
+    listVariableFields,
+    textVariableFields,
+  };
+}
+
+type VariableEditorState = ReturnType<typeof getInitialState>;
+
+function getVariableDefinitionFromState(state: VariableEditorState): VariableDefinition {
+  const { name, label, kind } = state;
+
+  const commonSpec = {
+    name,
+    display: {
+      label,
+    },
+  };
+
+  if (kind === 'TextVariable') {
+    return {
+      kind,
+      spec: {
+        ...commonSpec,
+        ...state.textVariableFields,
+      },
+    } as TextVariableDefinition;
+  }
+
+  if (kind === 'ListVariable') {
+    return {
+      kind,
+      spec: {
+        ...commonSpec,
+        allow_multiple: state.listVariableFields.allowMultiple,
+        allow_all_value: state.listVariableFields.allowAll,
+        plugin: state.listVariableFields.plugin,
+      },
+    } as ListVariableDefinition;
+  }
+  throw new Error(`Unknown variable kind: ${kind}`);
+}
+
+const SectionHeader = ({ children }: React.PropsWithChildren) => (
+  <Typography pb={2} variant="subtitle1">
+    {children}
+  </Typography>
+);
+
+export function VariableEditForm({
+  initialVariableDefinition,
+  onChange,
+  onCancel,
+}: {
+  initialVariableDefinition: VariableDefinition;
+  onChange: (def: VariableDefinition) => void;
+  onCancel: () => void;
+}) {
+  const { data: pluginList } = useListPluginMetadata('Variable');
+  const { getPlugin } = usePluginRegistry();
+  const [state, setState] = useImmer(getInitialState(initialVariableDefinition));
+
+  const updatePluginKind = async (kind: string) => {
+    const p = await getPlugin('Variable', kind);
+    setState((draft) => {
+      draft.listVariableFields.plugin.kind = kind;
+      draft.listVariableFields.plugin.spec = p.createInitialOptions();
+    });
+  };
```

Comment:
Hopefully (maybe?) the components I added in #624 might get rid of some of this logic (or handle it for you). In particular [`PluginEditor`](https://github.com/perses/perses/blob/2db8a47e34b744ceea450c4ebc9fc1c97e3a79cc/ui/plugin-system/src/components/PluginEditor/PluginEditor.tsx) or if you want the behavior with a slightly different UI layout the [`usePluginEditor`](https://github.com/perses/perses/blob/2db8a47e34b744ceea450c4ebc9fc1c97e3a79cc/ui/plugin-system/src/components/PluginEditor/plugin-editor-api.ts) hook might be what you want.

Reply:
I'll take a look at those and see if they make sense to integrate

Reply:
Integrated PluginEditor, it was pretty helpful. Had to make a few tweaks to styling, let me know if you think thats ok

Reply:
Totally cool on styling tweaks, I just basically tried to get something functional in there.